src/OSPSuite.Core/Domain/Services/ConcurrencyManager.cs

```diff
@@ -38,96 +36,76 @@ public interface IConcurrencyManager
       /// </summary>
       /// <typeparam name="TData">Data type to consume by the worker function</typeparam>
       /// <typeparam name="TResult">Data produced by the worker function</typeparam>
-      /// <param name="numberOfCoresToUse">Number of cores to use. Use 0 or negative to take all cores</param>
       /// <param name="data">List of data to consume by the workers</param>
-      /// <param name="idFunc"></param>
       /// <param name="action">
       ///    A function running on each worker on each piece of data. It is assume that the action runs on its
       ///    own thread
       /// </param>
       /// <param name="cancellationToken">Cancellation token to cancel the threads</param>
+      /// <param name="results">A concurrent dictionary holding the results after execution</param>
+      /// <param name="numberOfCoresToUse">Number of cores to use. Use 0 or negative to take all cores</param>
       /// <returns>Dictionary binding a result for each input data after running the action on it</returns>
-      Task<IReadOnlyDictionary<TData, ConcurrencyManagerResult<TResult>>> RunAsync<TData, TResult>(
-         int numberOfCoresToUse,
+      Task RunAsync<TData, TResult>(
          IReadOnlyList<TData> data,
-         Func<TData, string> idFunc,
-         Func<int, TData, CancellationToken, Task<TResult>> action,
-         CancellationToken cancellationToken
-      );
+         Func<TData, CancellationToken, TResult> action,
+         CancellationToken cancellationToken,
+         ConcurrentDictionary<TData, ConcurrencyManagerResult<TResult>> results,
+         int numberOfCoresToUse = 0
+      ) where TData : IWithId;
    }
 
    public class ConcurrencyManager : IConcurrencyManager
    {
-      private readonly int _maximumNumberOfCoresToUse = Math.Max(1, Environment.ProcessorCount - 1);
+      private readonly int _maximumNumberOfCoresToUse;
+
+      public ConcurrencyManager(ICoreUserSettings coreUserSettings)
+      {
+         _maximumNumberOfCoresToUse = Math.Max(1, coreUserSettings.MaximumNumberOfCoresToUse);
+      }
 
-      public async Task<IReadOnlyDictionary<TData, ConcurrencyManagerResult<TResult>>> RunAsync<TData, TResult>
-      (int numberOfCoresToUse,
+      public async Task RunAsync<TData, TResult>
+      (
          IReadOnlyList<TData> data,
-         Func<TData, string> idFunc,
-         Func<int, TData, CancellationToken, Task<TResult>> action,
-         CancellationToken cancellationToken)
+         Func<TData, CancellationToken, TResult> action,
+         CancellationToken cancellationToken,
+         ConcurrentDictionary<TData, ConcurrencyManagerResult<TResult>> results,
```

Comment:
Hhere the results are crfeated by the caller. Why? Let's have the task return the results no?

Reply:
I think there was a problem when throwing exceptions. If an exception is thrown then the whole list was lost. I rather think there was a bad exception handling so let us try again and see if it goes good with these changes.